File: BOJ/♠ 2000~2999/2805_나무_자르기.py
import sys
input = sys.stdin.readline
num, need = list(map(int, input().split()))
tree = list(map(int, input().split()))
start, end = 0, max(tree)

while start <= end:
    mid = (start + end) // 2
    sum_ = 0
    for i in range(num):
        if tree[i] > mid:
            sum_ += tree[i] - mid
    if sum_ >= need:
        start = mid + 1
    else:
        end = mid - 1

print(end)

File: BOJ/♠ 2000~2999/2805_나무_자르기.js
const isPossible = (height) => {
  let totalTreeHeight = 0;

  trees.forEach((tree) => {
    totalTreeHeight += Math.max(tree - height, 0);
  });

  return totalTreeHeight >= need;
}

const getAnswer = () => {
  let start = 0;
  let end = 2_000_000_000;
  let mid;

  while (start <= end) {
    mid = Math.floor((start + end) / 2);

    if (isPossible(mid)) {
      start = mid + 1;
    } else {
      end = mid - 1;
    }
  }

  return end;
}

const input = require('fs')
  .readFileSync(0, 'utf-8')
  .trim()
  .split('\n');

const need = Number(input[0].split(' ')[1]);
const trees = input[1].split(' ').map(Number);

console.log(getAnswer());